src/plot-netperfmeter-results.1
.\" ==========================================================================
.\"         _   _      _   ____            __ __  __      _
.\"        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
.\"        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
.\"        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
.\"        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
.\"
.\"                  NetPerfMeter -- Network Performance Meter
.\"                 Copyright (C) 2009-2026 by Thomas Dreibholz
.\" ==========================================================================
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact:  dreibh@simula.no
.\" Homepage: https://www.nntb.no/~dreibh/netperfmeter/
.\"
.\" ###### Setup ############################################################
.Dd May 27, 2026
.Dt plot\-netperfmeter\-results 1
.Os plot\-netperfmeter\-results
.\" ###### Name #############################################################
.Sh NAME
.Nm plot\-netperfmeter\-results
.Nd Network Performance Meter
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm plot\-netperfmeter\-results Ar config\_name
.br
.Op Fl w | Fl \-ownfile
.br
.Op Fl p | Fl \-per\-flow\-plots
.Nm plot\-netperfmeter\-results
.Op Fl h | Fl \-help
.Nm plot\-netperfmeter\-results
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm plot\-netperfmeter\-results
uses GNU R to plot measurement results written by NetPerfMeter into a PDF file. Optionally, a separate PDF file can be written for each page.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments have to be provided:
.Bl -tag -width indent
.It Ar config\_name
The name of the config file written by NetPerfMeter.
.It Fl w | Fl \-ownfile
Plot a separate PDF file for each page.
.It Fl p | Fl \-per\-flow\-plots
Also generate separate plots for each flow.
.It Fl h | Fl \-help
Prints command\-line parameters.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
Some plot\-netperfmeter\-results usage examples:
.Bl -tag -width indent
.It plot\-netperfmeter\-results output.config
Plot the results given by output.config. The resulting PDF file will be named output.pdf.
.It plot\-netperfmeter\-results output.config \--per\-flow\-plots
Plot the results given by output.config, including per\-flow plots. The resulting PDF file will be named output.pdf.
.It plot\-netperfmeter\-results output.config \--ownfile
Plot the results given by output.config. For each page, a separate PDF file is written. The resulting PDF files will be named output\-<title>.pdf, where <title> corresponds to the page title.
.It plot\-netperfmeter\-results \--version
.It plot\-netperfmeter\-results \--help
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr netperfmeter 1 ,
.Xr pdfembedfonts 1 ,
.Xr setpdfmetadata 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of NetPerfMeter. The latest version of NetPerfMeter can be found on the NetPerfMeter Homepage at
.Lk https://www.nntb.no/\(tidreibh/netperfmeter/ "NetPerfMeter Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 1f383a60d2d770535581e46f78444bcf9c4f4ebf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 11:29:02 2026 +0200

    * Make inclusion of examples (scripts, results) optional in CMakeLists.txt.
    * Moved obsolete test program.

commit d26562086455bbf41d21f6c3cae99b6cae1bf2a4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:43:28 2026 +0200

    Improved FreeBSD port.

commit 946bfafbadddcea276a7aa360c8ee697695c69db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:42:01 2026 +0200

    Fixed includes.

commit 9d186d0746f498d32427611d54b5a927e5baa773
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 09:58:58 2026 +0200

    RPM spec fix.

commit 26e1ca4d015ec9403a22df90b8600605839e44e7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:13:55 2026 +0200

    New release netperfmeter-2.0.4.

commit 02576769d8562e11e6b0cb7d15200fe7d40e86cf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:12:54 2026 +0200

    Minor manpage update.

commit bf272b25bfa5094a78b4657f4bd9912dcb23edb2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:03:05 2026 +0200

    Minor clean-up.

commit b732393412bd54a0e4d0f1347ede6747b7a7a870
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:01:02 2026 +0200

    Added option to configure the separator.

commit 977df2a565b3cf11fba2e83e8ad00f11c8c570ec
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:53:07 2026 +0200

    Small clean-ups.

commit 862fb32f3d7abcac2fff6537fba36b34981d9ffa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:35:21 2026 +0200

    Further clean-ups.

commit 1d6d37a6dfac84a1858845f33fd9df66ff1d4806
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:31:15 2026 +0200

    Further clean-ups.

commit 793d87f6b1d0746c09488178c0e84424ce544115
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:30:39 2026 +0200

    Added option to configure the separator.

commit 2148af89133efdc3beeffc3f0c4a090e9cb75468
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:28:24 2026 +0200

    Improved shell completions.

commit 336be30bc58436d386d92183522a263cc30d1fd7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:22:41 2026 +0200

    Improved vector counting.

commit 8a5797d64ac46d71c275374869167aa5170cec12
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:19:01 2026 +0200

    Further improvements.

commit eb7c8be6269e2ca1790a0ec099201b3575516cc9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 13:33:00 2026 +0200

    Code improvements.

commit 05228c17745801c8910c7d06570f9ab804a20215
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:13:34 2026 +0200

    CMakeLists.txt update.

commit 1b427da57b4c7563f2418755c0439f4b8b7b076f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:08:58 2026 +0200

    Added more shell completions.

commit b799cc0179b7f960c9ef4d1fb28c5a9b34794e43
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:51:22 2026 +0200

    Options handling improvements.

commit 69b965905364e4306d1a8ef6149aa879136fca84
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:45:50 2026 +0200

    Packaging updates.

commit 82e1980494ab0fc8731b52671794c3c52a736c6b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:44:06 2026 +0200

    Added shell completions.

commit 8181abb986847144955c1e0b6e8da2c6dc863125
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:17:37 2026 +0200

    Further manpage improvements.

commit 3037cb9b5c9417375a288df233949c2e2ce9431a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:06:07 2026 +0200

    Improved manpages.

commit 428bdbaa9bccd79ed23d82f2a02157974ff70c51
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:57 2026 +0200

    Added version option.

commit 5eb240346cab65c110a07b63b0a2e4d748898b97
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:40 2026 +0200

    Added version option.

commit ef9d7f39317e1017354170936b4d5a359dce0f0e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 22:31:03 2026 +0200

    Improved CMakeLists.txt.

commit 58524abe9716d4f1a4f092f5aa64faec49fce966
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 21:03:35 2026 +0200

    Further manpage improvements.

commit 8c2c8ef1307836b855719897b2bcdb4775c11abc
Merge: 68c78ee b2d981b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 20:54:04 2026 +0200

    Merge branch 'master' of github.com:dreibh/netperfmeter

commit b2d981b677bc1bce0f2a9772a7905dff05375d3e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 16:46:28 2026 +0200

    Improved manpages.

commit dadcd3763d363a0a401d64439996f24eefe8780a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 25 22:37:19 2026 +0200

    A quick compile test under MacOS. Some more work would be needed to work around missing SCTP.

commit 68c78ee1c78dd0dc42e1131bae4064f039b1d7c3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 9 10:36:06 2026 +0200

    Minor shell completion clean-up.

commit dcde13391c1f5dd16ce7f6db0da401713a3f310a
Author: Thomas Dreibholz <dreibh@simula.no>

src/plot-netperfmeter-results.R
#!/usr/bin/env Rscript
# ==========================================================================
#         _   _      _   ____            __ __  __      _
#        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
#        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
#        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
#        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
#
#                  NetPerfMeter -- Network Performance Meter
#                 Copyright (C) 2009-2026 by Thomas Dreibholz
# ==========================================================================
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Contact:  dreibh@simula.no
# Homepage: https://www.nntb.no/~dreibh/netperfmeter/

library("data.table", warn.conflicts = FALSE)
library("dplyr",      warn.conflicts = FALSE)
library("ggplot2")


# ###########################################################################
# #### PDF Metadata Handling                                             ####
# ###########################################################################

pdfMetadataFile <- NA
pdfMetadataPage <- NA


# ###### Set global variable (given as name string) to given value ##########
setGlobalVariable <- function(variable, value)
{
   globalEnv <- sys.frame()
   assign(variable, value, envir=globalEnv)
}


# ###### Create PDF info ####################################################
openPDFMetadata <- function(name)
{
   setGlobalVariable("pdfMetadataFile", file(paste(sep="", name, ".meta"), "w"))
   setGlobalVariable("pdfMetadataPage", 0)

   cat(sep="", "title NetPerfMeter Results Plots\n", file=pdfMetadataFile);
   cat(sep="", "subject Measurement Results\n", file=pdfMetadataFile);
   cat(sep="", "creator plot-netperfmeter-results\n", file=pdfMetadataFile);
   cat(sep="", "producer GNU R/Ghostscript\n", file=pdfMetadataFile);
   cat(sep="", "keywords: NetPerfMeter, Measurements, Results\n", file=pdfMetadataFile);
}


# ###### Finish current page and increase page counter ######################
nextPageInPDFMetadata <- function()
{
   setGlobalVariable("pdfMetadataPage", pdfMetadataPage + 1)
}


# ###### Add entry into PDF info ############################################
addBookmarkInPDFMetadata <- function(level, title)
{
   if(!is.na(pdfMetadataFile)) {
      cat(sep="", "outline ", level, " ", pdfMetadataPage + 1, " ", title, "\n", file=pdfMetadataFile)
    }
}


# ###### Close PDF info #####################################################
closePDFMetadata <- function()
{
   if(!is.na(pdfMetadataFile)) {
      close(pdfMetadataFile)
      setGlobalVariable("pdfMetadataFile", NA)
   }
   if(!is.na(pdfMetadataFile)) {
      close(pdfMetadataFile)
      setGlobalVariable("pdfMetadataFile", NA)
   }
}



# ###########################################################################
# #### Plotting Themes                                                   ####
# ###########################################################################

# ###### Plot themes ########################################################
PLOTTHEME_PAPER <- theme(
   legend.position      = "bottom",
   legend.justification = "center",
   legend.title         = element_text(size=12, face = "bold", colour="black", hjust=0.5),
   legend.text          = element_text(size=12),
   legend.key.size      = unit(1.00, "cm"),
   legend.key.width     = unit(1.50, "cm"),
   legend.background    = element_blank(),

   plot.title           = element_blank(),
   axis.title           = element_text(size=12, face="bold"),
   axis.text.x          = element_text(size=10, colour="black"),
   axis.text.y          = element_text(size=10, angle=90, hjust=0.5, colour="black"),
   axis.ticks           = element_line(colour = "black"),

   strip.text           = element_text(size=10, face="bold"),

   panel.border         = element_rect(colour = "black", fill = NA),
   panel.background     = element_blank(),
   panel.grid.major     = element_line(color = "gray40", linewidth = 0.25, linetype = 1),
   panel.grid.minor     = element_line(color = "gray20", linewidth = 0.10, linetype = 2)
)

PLOTTHEME_FANCY <- theme(
   legend.position      = "bottom",
   legend.justification = "center",
   legend.title         = element_text(size=12, face = "bold", colour="black", hjust=0.5),
   legend.text          = element_text(size=12),
   legend.key.size      = unit(1.00, "cm"),
   legend.key.width     = unit(1.50, "cm"),
   # legend.background    = element_rect(colour = "black", fill = "#ffffaa55", linewidth=1),
   # legend.background    = element_rect(colour = "black", fill = NA, linewidth=1),
   legend.background    = element_blank(),

   plot.title           = element_text(size=16, hjust = 0.5, face="bold"),
   axis.title           = element_text(size=12, face="bold"),
   axis.text.x          = element_text(size=10, colour="black"),
   axis.text.y          = element_text(size=10, angle=90, hjust=0.5, colour="black"),
   axis.ticks           = element_line(colour = "black"),

   strip.text           = element_text(size=10, face="bold"),

   panel.border         = element_rect(colour = "black", fill = NA),
   # panel.background     = element_blank(),
   panel.grid.major     = element_line(color = "gray40", linewidth = 0.25, linetype = 1),
   panel.grid.minor     = element_line(color = "gray20", linewidth = 0.10, linetype = 2)
)



# ###########################################################################
# #### Plotting                                                          ####
# ###########################################################################

# ###### Create a plot ######################################################
createPlot <- function(dataSet, title, ySet, yTitle, baseColor, zSet, zTitle,
                       vSet = c(), vTitle = "", bookmarkLevel = 3)
{
   cat(sep="", pdfMetadataPage + 1, ": Plotting ", title, " ...\n")

   if(plotOwnFile) {
      names   <- unlist(strsplit(title, " "))
      pdfName <- paste(sep="", pdfFilePrefix, "-")
      for(i in 1:length(names)) {
         pdfName <- paste(sep="", pdfName, names[i])
      }
      pdfName <- paste(sep="", pdfName, ".pdf")
      cat(pdfName,"\n")
      pdf(pdfName, width=plotWidth, height=plotHeight, paper=plotPaper,
          family=plotFontFamily, pointsize=plotPointSize)
   }
   else {
      addBookmarkInPDFMetadata(bookmarkLevel, title)
   }

   p <- ggplot(dataSet,
               aes(x     = RelTime,
                   y     = ySet,
                   color = zSet,
                   linetype = vSet)) +
           PLOTTHEME_FANCY +
           theme(panel.border = element_rect(colour = baseColor, fill = NA, linewidth = 2)) +
           geom_line(linewidth = 1) +
           labs(title = title,
                x     = "Time t [s]",
                y     = yTitle,
                color = zTitle,
                linetype = vTitle) +
           guides(color    = guide_legend(ncol = 3),
                  linetype = guide_legend(ncol = 1))
   print(p)

   if(plotOwnFile) {
      dev.off()
      processPDFbyGhostscript(pdfName)
   }
   else {
      nextPageInPDFMetadata()
   }
}



# ###########################################################################
# #### Statistics                                                        ####
# ###########################################################################

# ###### Create node plots ##################################################
plotNodeStats <- function(inputData, nodeName)
{
   addBookmarkInPDFMetadata(1, paste(sep="", "Node ''", nodeName, "''"))
   data <- inputData %>% filter( Action != "Lost" )

   # ====== Input/Output Rates ==============================================
   addBookmarkInPDFMetadata(2, "Input/Output Rates")
   createPlot(data, paste(sep="", "Bit Rate Sent/Received on Node ''", nodeName, "''"),
              (data$RelBytes * 8 / 1000) / data$Interval, "Bit Rate [Kbit/s]", "blue4",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Byte Rate Sent/Received on Node ''", nodeName, "''"),
              (data$RelBytes / 1000) / data$Interval, "Byte Rate [KiB/s]", "blue2",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Packet Rate Sent/Received on Node ''", nodeName, "''"),
              data$RelPackets / data$Interval, "Packet Rate [Packets/s]", "green4",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Frame Rate Sent/Received on Node ''", nodeName, "''"),
              data$RelFrames / data$Interval, "Frame Rate [Frames/s]", "yellow4",
              data$Description, "Flow F", data$Action, "Action A")

   # ====== Input/Output Volumes ============================================
   addBookmarkInPDFMetadata(2, "Input/Output Volumes")
   createPlot(data, paste(sep="", "Bits Sent/Received on Node ''", nodeName, "''"),
              (data$AbsBytes * 8 / 1000), "Bits [Kbit]", "blue4",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Bytes Sent/Received on Node ''", nodeName, "''"),
              (data$AbsBytes / 1000), "Bytes [KiB]", "blue2",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Packets Sent/Received on Node ''", nodeName, "''"),
              data$AbsPackets, "Packets [1]", "green4",
              data$Description, "Flow F", data$Action, "Action A")
   createPlot(data, paste(sep="", "Frames Sent/Received on Node ''", nodeName, "''"),
              data$AbsFrames, "Frames [s]", "yellow4",
              data$Description, "Flow F", data$Action, "Action A")

   # ====== Jitter ==========================================================
   addBookmarkInPDFMetadata(2, "Quality of Service")
   data <- inputData %>% filter( Action == "Received" )

   createPlot(data, paste(sep="", "Jitter on Node ''", nodeName, "''"),
              data$Jitter, "Jitter [ms]", "gold4",
              data$Description, "Flow F", data$Action, "Action A")

   # ====== Loss ============================================================
   data <- inputData %>% filter( Action == "Lost" )
   createPlot(data, paste(sep="", "Bit Loss Rate on Node ''", nodeName, "''"),
              (data$RelBytes * 8 / 1000) / data$Interval, "Byte Loss Rate [Kbit/s]", "red4",
              data$Description, "Flow F")
   createPlot(data, paste(sep="", "Byte Loss Rate on Node ''", nodeName, "''"),
              data$RelBytes / data$Interval, "Byte Loss Rate [KiB/s]", "red2",
              data$Description, "Flow F")
   createPlot(data, paste(sep="", "Packet Loss Rate on Node ''", nodeName, "''"),
              data$RelPackets / data$Interval, "Packet Loss Rate [Packets/s]", "red3",
              data$Description, "Flow F")
   createPlot(data, paste(sep="", "Frame Loss Rate on Node ''", nodeName, "''"),
              data$RelFrames / data$Interval, "Frame Loss Rate [Frames/s]", "red1",
              data$Description, "Flow F")
}


# ###### Create flow QoS plots ##############################################
plotQoSStatistics <- function(inputData, flowSummaryData, nodeName) {
   addBookmarkInPDFMetadata(1, "Per-Flow Results")

   for(isActive in c(0, 1)) {
      if(isActive == 1) {
         nodeName = NAME_ACTIVE_NODE
      }
      else {
         nodeName <- NAME_PASSIVE_NODE
      }
      addBookmarkInPDFMetadata(2, paste(sep="", "Node ''", nodeName, "''"))

      flowLevels <- levels(factor(flowSummaryData$FlowID))
      for(i in seq(1, length(flowLevels))) {
         data <- subset(flowSummaryData, (flowSummaryData$FlowID == flowLevels[i]))
         flowNames <- levels(factor(data$FlowName))
         flowName <- flowNames[1]
         addBookmarkInPDFMetadata(3, paste(sep="", "Flow ''", flowName, "''"))


         # ====== Input/Output Rates ========================================
         addBookmarkInPDFMetadata(4, "Input/Output Rates")
         data <- subset(inputData,
                        (inputData$FlowID == flowLevels[i]) &
                           (inputData$IsActive == isActive) & (inputData$FlowID != -1) &
                           (inputData$Action != "Lost"))
         createPlot(data, paste(sep="", "Bit Rate Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    (data$RelBytes * 8 / 1000) / data$Interval, "Bit Rate [Kbit/s]", "blue4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Byte Rate Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    (data$RelBytes / 1000) / data$Interval, "Byte Rate [KiB/s]", "blue2",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Packet Rate Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    data$RelPackets / data$Interval, "Packet Rate [Packets/s]", "green4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Frame Rate Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    data$RelFrames / data$Interval, "Frame Rate [Frames/s]", "yellow4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)

         # ====== Input/Output Absolute =====================================
         addBookmarkInPDFMetadata(4, "Input/Output Absolute")
         data <- subset(inputData,
                        (inputData$FlowID == flowLevels[i]) &
                           (inputData$IsActive == isActive) & (inputData$FlowID != -1) )
                           # & (inputData$Action != "Lost"))
         createPlot(data, paste(sep="", "Bits Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    (data$AbsBytes * 8 / 1000), "Bits [Kbit]", "blue4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Bytes Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    (data$AbsBytes / 1000), "Bytes [KiB]", "blue2",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Packets Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    data$AbsPackets, "Packets [1]", "green4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)
         createPlot(data, paste(sep="", "Frames Sent/Received for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                    data$AbsFrames, "Frames [s]", "yellow4",
                    data$Action, "Action A", data$Description, "Flow F", bookmarkLevel=5)

         # ====== Quality of Service ========================================
         addBookmarkInPDFMetadata(4, "Quality of Service")
         data <- subset(flowSummaryData,
                        (flowSummaryData$FlowID == flowLevels[i]) &
                           (flowSummaryData$IsActive == isActive))
         if(length(data$FlowID) > 0) {   # May be empty, if no packets have been sent in this direction!
            createPlot(data, paste(sep="", "Per-Message Delay for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                       data$Delay, "Delay [ms]", "orange2",
                       c(), NA,   # data$NodeName, "Node{N}",
                       bookmarkLevel=5)
            createPlot(data, paste(sep="", "Per-Message Jitter for Flow ''", flowName, "'' on Node ''", nodeName, "''"),
                       data$Jitter, "Jitter [ms]", "gold4",
                       c(), NA,   # data$NodeName, "Node{N}",
                       bookmarkLevel=5)
         }
      }
   }
}



# ###########################################################################
# #### Main Program                                                      ####
# ###########################################################################

# ###### Default Settings ###################################################
plotOwnFile    <- TRUE
plotFontFamily <- "Helvetica"
plotPointSize  <- 12
plotWidth      <- -1
plotHeight     <- -1
plotPaper      <- "A4r"   # Use "special" for manual values! Or: A4/A4r.


# ###### Command-Line Arguments #############################################
args <- commandArgs(TRUE)
# print(args)
for(i in 1:length(args)) {
   eval(parse(text=args[i]))
}

# cat(sep="", "programDirectory=", programDirectory, "\n")
# cat(sep="", "configFile=",       configFile,       "\n")
# cat(sep="", "summaryFile=",      summaryFile,      "\n")
# cat(sep="", "pdfFilePrefix=",    pdfFilePrefix,    "\n")
# cat(sep="", "perFlowPlots=",     perFlowPlots,     "\n")
# cat(sep="", "plotOwnFile=",      plotOwnFile,      "\n")


# ====== Load input data ====================================================
source(configFile)
summaryData <- data.table(read.csv(summaryFile, sep="\t")) %>%
                  filter(Interval > 0)   # Avoids "divide by zero" on first entry
flowSummaryData <- data.table(read.csv(flowSummaryFile, sep="\t"))

setGlobalVariable("gRuntime", round(max(summaryData$RelTime)))
cat(sep="", "Runtime=", gRuntime, "\n")


# ====== Begin writing PDF file =============================================
if(!plotOwnFile) {
   pdfFileName <- paste(sep="", pdfFilePrefix, "-TEMP.pdf")
   pdf(pdfFileName, width=plotWidth, height=plotHeight, paper=plotPaper,
       family=plotFontFamily, pointsize=plotPointSize)
   openPDFMetadata(pdfFileName)
}


# ====== Create plots for active and passive node ===========================
data <- summaryData %>% filter( IsActive == 1, FlowID != -1 )
plotNodeStats(data, NAME_ACTIVE_NODE)

data <- summaryData %>% filter( IsActive == 0, FlowID != -1 )
plotNodeStats(data, NAME_PASSIVE_NODE)


# ====== Flow QoS Statistics ================================================
if(perFlowPlots == 1) {
   plotQoSStatistics(summaryData, flowSummaryData)
}


# ====== Finish PDF file ====================================================
if(!plotOwnFile) {
   # ------ Close PDF and embed fonts (using Ghostscript) -------------------
   dev.off()
   closePDFMetadata()

   # ------ Add PDF outlines and meta data ----------------------------------
   cmd1 <- paste(sep="", "setpdfmetadata ", pdfFileName, " ", pdfFileName, ".meta ", pdfFilePrefix, "-TEMP2.pdf",
                         " || mv ", pdfFileName, " ", pdfFilePrefix, "-TEMP2.pdf")
   # cat(cmd1,"\n")
   ret1 <- system(cmd1)
   if(ret1 != 0) {
      stop(gettextf("status %d in running command '%s'", ret1, cmd1))
   }

   # ------ Add PDF outlines and metadata -----------------------------------
   cmd2 <- paste(sep="", "pdfembedfonts ", pdfFilePrefix, "-TEMP2.pdf", " ", pdfFilePrefix, ".pdf -optimize",
                         " || mv ", pdfFilePrefix, "-TEMP2.pdf", " ", pdfFilePrefix, ".pdf")
   # cat(cmd2,"\n")
   ret2 <- system(cmd2)
   if(ret2 != 0) {
      stop(gettextf("status %d in running command '%s'", ret2, cmd2))
   }

   # ------ Remove temporary files ------------------------------------------
   unlink(pdfFileName)
   unlink(paste(sep="", pdfFilePrefix, "-TEMP2.pdf"))
   unlink(paste(sep="", pdfFileName, ".meta"))
}
quit(runLast=FALSE)
